src/core/toga/sources/stack_source.py

```diff
@@ -0,0 +1,44 @@
+from .collection_source import CollectionSource
+from .row import Row
+
+
+class StackSource(CollectionSource):
+    """A data source to store a stack of multiple data values, in a row-like fashion.
+
+    Support pushing and popping rows and uniqueness
+    """
+
+    def __init__(self, accessors, unique=True, size=None):
+        super(StackSource, self).__init__(data=[], accessors=accessors)
+        self.unique = unique
+        self.size = size
+
+    def push(self, *values, **named):
+        row = Row.create_row(
+            data=dict(zip(self._accessors, values), **named),
+            accessors=self._accessors,
+            source=self,
+        )
+        if self.unique:
+            try:
+                existing_index = next(
+                    filter(
+                        lambda index_row_tuple: index_row_tuple[1] == row,
+                        enumerate(self._data)
```

Comment:
Iterators are a bit more expensive than simple iteration; a simple while loop would likely also be easier for future explorers to understand.